prefect_sqlalchemy/credentials.py

```diff
@@ -225,31 +343,400 @@ def sqlalchemy_credentials_flow():
-            self.rendered_url.render_as_string(hide_password=False)
+        # can't use run_sync_in_worker_thread:
+        # ProgrammingError: (sqlite3.ProgrammingError) SQLite objects created in a
+        # thread can only be used in that same thread.
+        result_set = connection.execute(*execute_args, **execute_kwargs)
+
+        if self._driver_is_async:
+            result_set = await result_set
+            await connection.commit()  # very important
+        return result_set
+
+    @asynccontextmanager
+    async def _manage_connection(self, **get_client_kwargs: Dict[str, Any]):
+        if self._driver_is_async:
+            async with self.get_client(**get_client_kwargs) as connection:
+                yield connection
+        else:
+            with self.get_client(**get_client_kwargs) as connection:
+                yield connection
+
+    async def _get_result_set(
+        self, *execute_args: Tuple[Any], **execute_kwargs: Dict[str, Any]
+    ) -> CursorResult:
+        """
+        Returns a new or existing result set based on whether the inputs
+        are unique.
+
+        Args:
+            *execute_args: Args to pass to execute.
+            **execute_kwargs: Keyword args to pass to execute.
+
+        Returns:
+            The result set from the operation.
+        """  # noqa: E501
+        input_hash = hash_objects(*execute_args, **execute_kwargs)
+        assert input_hash is not None, (
+            "We were not able to hash your inputs, "
+            "which resulted in an unexpected data return; "
+            "please open an issue with a reproducible example."
         )
-        return d
+
+        if input_hash not in self._unique_results.keys():
+            if self._driver_is_async:
+                connection = await self._exit_stack.enter_async_context(
+                    self.get_client()
+                )
+            else:
+                connection = self._exit_stack.enter_context(self.get_client())
+            result_set = await self._async_sync_execute(
+                connection, *execute_args, **execute_kwargs
+            )
+            # implicitly store the connection by storing the result set
+            # which points to its parent connection
+            self._unique_results[input_hash] = result_set
+        else:
+            result_set = self._unique_results[input_hash]
+        return result_set
+
+    def _reset_cursor_results(self) -> None:
+        """
+        Closes all the existing cursor results.
+        """
+        input_hashes = tuple(self._unique_results.keys())
+        for input_hash in input_hashes:
+            try:
+                cursor_result = self._unique_results.pop(input_hash)
+                cursor_result.close()
+            except Exception as exc:
+                self.logger.warning(
+                    f"Failed to close connection for input hash {input_hash!r}: {exc}"
+                )
+
+    def reset_connections(self) -> None:
```

Comment:
Couldn't just make this into a single sync_compatible method because `__exit__` depends on this through `close`, and when that's not sync_compatible, it breaks the chain of sync compatibility.